### DISCUSSION TITLE
Body parameter cannot be a pydantic union anymore since v0.95

### DISCUSSION THREAD

**Discussion by morian** [fastapi/fastapi#9287] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Annotated, Literal

from fastapi import FastAPI
from pydantic import BaseModel, Field


class ItemV1(BaseModel):
    version: Literal[1] = 1

class ItemV2(BaseModel):
    version: Literal[2] = 2

Item = Annotated[
    ItemV1 | ItemV2,
    Field(discriminator='version'),
]

app = FastAPI()

@app.post('/')
async def create(body: Item) -> None:
    pass
```

### Description

* Run this code with `uvicorn 'main:app'`

It raises the following backtrace:
```
Traceback (most recent call last):
[...]
  File "/tmp/lab/main.py", line 20, in <module>
    @app.post('/')
     ^^^^^^^^^^^^^
  File "/tmp/lab/venv/lib/python3.11/site-packages/fastapi/routing.py", line 661, in decorator
    self.add_api_route(
  File "/tmp/lab/venv/lib/python3.11/site-packages/fastapi/routing.py", line 600, in add_api_route
    route = route_class(
            ^^^^^^^^^^^^
  File "/tmp/lab/venv/lib/python3.11/site-packages/fastapi/routing.py", line 446, in __init__
    self.dependant = get_dependant(path=self.path_format, call=self.endpoint)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/lab/venv/lib/python3.11/site-packages/fastapi/dependencies/utils.py", line 328, in get_dependant
    if is_body_param(param_field=param_field, is_path_param=is_path_param):
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/lab/venv/lib/python3.11/site-packages/fastapi/dependencies/utils.py", line 491, in is_body_param
    assert isinstance(
AssertionError: Param: body can only be a request body, using Body()
```

The same code used to work well with FastAPI v0.94.1 (and priors).
I tried to annotate using `Annotated[Item, Body()]` but it did not fix anything:
```
Traceback (most recent call last):
[...]
  File "/tmp/lab/main.py", line 20, in <module>
    @app.post('/')
     ^^^^^^^^^^^^^
  File "/tmp/lab/venv/lib/python3.11/site-packages/fastapi/routing.py", line 661, in decorator
    self.add_api_route(
  File "/tmp/lab/venv/lib/python3.11/site-packages/fastapi/routing.py", line 600, in add_api_route
    route = route_class(
            ^^^^^^^^^^^^
  File "/tmp/lab/venv/lib/python3.11/site-packages/fastapi/routing.py", line 446, in __init__
    self.dependant = get_dependant(path=self.path_format, call=self.endpoint)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/lab/venv/lib/python3.11/site-packages/fastapi/dependencies/utils.py", line 303, in get_dependant
    type_annotation, depends, param_field = analyze_param(
                                            ^^^^^^^^^^^^^^
  File "/tmp/lab/venv/lib/python3.11/site-packages/fastapi/dependencies/utils.py", line 382, in analyze_param
    len(fastapi_annotations) <= 1
AssertionError: Cannot specify multiple `Annotated` FastAPI arguments for 'body'
```

Please let me know if there is any additional inputs I can provide or whether this is now a desired behavior.
Thanks!

### Operating System

Linux

### FastAPI Version

0.95.0

### Python Version

Python 3.11.2

### Additional Context

Pip freeze:
```
anyio==3.6.2
click==8.1.3
fastapi==0.95.0
h11==0.14.0
httptools==0.5.0
idna==3.4
pydantic==1.10.6
python-dotenv==1.0.0
PyYAML==6.0
sniffio==1.3.0
starlette==0.26.1
typing_extensions==4.5.0
uvicorn==0.21.1
uvloop==0.17.0
watchfiles==0.18.1
websockets==10.4
```

**Comment by phillipuniverse** [comment]
Same issue here, certainly seems like a bug in the recent 0.95.0.

**Reply by phillipuniverse** [reply]
Appears to be caused by the changes at https://github.com/tiangolo/fastapi/pull/4871

**Comment by stephen-christie-ss** [comment]
I also experienced this issue, updating to 0.95.0.

**Comment by Lawouach** [comment]
I got bitten today as well.

**Comment by leokhachatorians** [comment]
So uh, when does a 'discussion' become an 'issue'? Or are we just yelling into the void

**Comment by nzig** [comment, marked as answer]
This was indeed caused by the changes in #4871. Apologies for the inconvenience.

The quick solution to this problem is to replace
```python
Item = Annotated[
    ItemV1 | ItemV2,
    Field(discriminator='version'),
]
```
with
```python
Item = Annotated[
    ItemV1 | ItemV2,
    Body(discriminator='version'),
]
```

I'm not really sure whether the original code was supposed to work before, or if it just worked incidentally. 

@tiangolo what do you think? If you think it should work, I can try and figure out how to make it work.

**Reply by phillipuniverse** [reply]
@nzig thanks so much for the workaround! I just tried it in my project and it works perfectly.

In fact, it works even better than I thought it would, maybe the original `Field(discriminator...` wasn't supposed to work before as you said. You can see here in this OpenAPI diff after using `Body(discriminator...)`:

<img width="548" alt="image" src="https://user-images.githubusercontent.com/684275/230603417-4fcff766-4c45-4543-a7ef-50d3d4229f0a.png">

Before, the union had generated an `anyOf`, but now it generates the correct discriminated union with `oneOf`. I clearly missed this before, `oneOf` is the correct form.

The difference here maybe is a subtletly between Pydantic schema definitions and FastAPI schema definitions. The disadvantage of forcing `Body` in this annotation is that it makes it more difficult to apply a Pydantic-only concept (`Field(discriminator...)`) to the FastAPI context.

**Reply by karolpawlowski** [reply]
Thanks @nzig and @phillipuniverse . What do you suggest to do with pydantic unions that aren't used as a Body parameter but rather as a returned object type? Keep using `Field(discriminator=...)` or change it to Body or something else?

**Reply by nzig** [reply]
I think either will work, because `Body` is a subclass of the `FieldInfo` type that `Field()` returns.

I would use `Field()` unless the type definition is shared with something else that requires `Body`.

**Reply by karolpawlowski** [reply]
Thanks @nzig !

**Reply by stdkoehler** [reply]
This also bit me. Thanks for providing a solution.

If I might add my view on the matter: Personally, I would prefer if we could use Field as was the case in the previous FastAPI versions.
1) **Separation of responsibility**: When defining the models, we are in the "pydantic world" (until now I did not have any FastAPI dependency in the file where I defined my models). Using those models would be the responsibility of FastAPI, mixing FastAPI into the models themselves would be no good practice (imho)
2) **pydantic standard** A _Discriminated Union_ is defined by pydantic to create an OpenAPI compliant structure (https://docs.pydantic.dev/latest/usage/types/#discriminated-unions-aka-tagged-unions). Here explicitly the use of _Field_ is shown (and still works when nested into models. Only when providing the Discriminated Union directly as Request Body do we trigger the "bug")
3) **Ease of Use / Documentation** The pydantic doc is where probably most people will land, when they are looking on how to get Discriminators into FastAPI. There is no according documentation (at least I found none) how to do it in FastAPI. (But mostly you just need to know pydantic so that's fine - **_except now: this case_**). From my research, this discussion thread is the only place where you could find that you need to use Body now. (And I didn't find in via search engine - I had to click through 10 pages of discussion looking into every thread whose description could've been related to my problem)

**Reply by fortzi** [reply]
Thanks for your support @nzig, really appreciated!

Unfortunately the suggested solution doesn't work for me, as I'm using an external package and cannot control its typing - I cannot change the `Annotated` type from `Field` to `Body`. Hence I cannot update FastAPI, which is also blocking me from using pydantic2...

Is this issue expected to be addressed? I'm sure that many more can't use the suggested workaround (or won't due to the reasons that @stdkoehler laid out)

**Reply by nzig** [reply]
If you're migrating to pydantic v2, could wrapping the third-party type in a [`RootModel`](https://docs.pydantic.dev/latest/usage/models/#rootmodel-and-custom-root-types) solve your problem?

**Comment by pr0gg3d** [comment]
Hi, first of all thanks to FastAPI developers for such a great project!

Just wanted to know if there's any progress on this, because it seems that it's a showstopper for fully migrating to pydantic V2.

I also agree with a solution that allows the use of `Field(discriminator=....)`, for the reasons that @stdkoehler pointed out very well.

For instance, in our projects we are using these Discriminated Unions throughout the application logic, where FastAPI it's only the final API part in order to interact with outside world.

Having to mantain 2 identical sets of Discriminated Unions, only by changing Field() into Body() if these are going to be used in FastAPI is not efficient at all.

Thanks again!

**Comment by psychedelicious** [comment]
@pr0gg3d This issue is resolved for us by #10773

**Reply by morian** [reply]
Indeed, it also now works for me as well. Closing this discussion.
